openedx/core/lib/gating/api.py

```diff
@@ -0,0 +1,306 @@
+        block_ids.append(block_id)
+        milestones_by_block_id[block_id] = milestone
+
+    result = []
+    for block in modulestore().get_items(course_key, qualifiers={'name': block_ids}):
+        milestone = milestones_by_block_id.get(block.location.block_id)
+        if milestone:
+            milestone['block_display_name'] = block.display_name
+            milestone['block_usage_key'] = unicode(block.location)
+            result.append(milestone)
+
+    return result
+
+
+@milestones_active()
+def add_prerequisite(course_key, prereq_content_key):
+    """
+    Creates a new Milestone and CourseContentMilestone indicating that
+    the given course content fulfills a prerequisite for gating
+
+    Arguments:
+        course_key (str|CourseKey): The course key
+        prereq_content_key (str|UsageKey): The prerequisite content usage key
+
+    Returns:
+        None
+    """
+    milestone = milestones_api.add_milestone(
+        {
+            'name': _('Gating milestone for {usage_key}').format(usage_key=unicode(prereq_content_key)),
+            'namespace': "{usage_key}{qualifier}".format(
+                usage_key=prereq_content_key,
+                qualifier=GATING_NAMESPACE_QUALIFIER
+            ),
+            'description': _('System defined milestone'),
+        },
+        propagate=False
+    )
+    milestones_api.add_course_content_milestone(course_key, prereq_content_key, 'fulfills', milestone)
+
+
+@milestones_active()
+def remove_prerequisite(prereq_content_key):
+    """
+    Removes the Milestone and CourseContentMilestones related to the gating
+    prerequisite which the given course content fulfills
+
+    Arguments:
+        prereq_content_key (str|UsageKey): The prerequisite content usage key
+
+    Returns:
+        None
+    """
+    milestones = milestones_api.get_milestones("{usage_key}{qualifier}".format(
+        usage_key=prereq_content_key,
+        qualifier=GATING_NAMESPACE_QUALIFIER
+    ))
+    for milestone in milestones:
+        milestones_api.remove_milestone(milestone.get('id'))
+
+
+@milestones_active(default=False)
+def is_prerequisite(course_key, prereq_content_key):
+    """
+    Returns True if there is at least one CourseContentMilestone
+    which the given course content fulfills
+
+    Arguments:
+        course_key (str|CourseKey): The course key
+        prereq_content_key (str|UsageKey): The prerequisite content usage key
+
+    Returns:
+        bool: True if the course content fulfills a CourseContentMilestone, otherwise False
+    """
+    return get_gating_milestone(
+        course_key,
+        prereq_content_key,
+        'fulfills'
+    ) is not None
+
+
+@milestones_active()
+def set_required_content(course_key, gated_content_key, prereq_content_key, min_score):
+    """
+    Adds a `requires` milestone relationship for the given gated_content_key if a prerequisite
+    prereq_content_key is provided. If prereq_content_key is None, removes the `requires`
```

Comment:
I don't see that the code removes the requires relationship if prereq_content_key is None (which seems like a bad idea to me anyway, but the doc should be updated).